### PR TITLE
Add prelimary compositor interface

### PR DESCRIPTION
**Pull request by neithernut**
This PR adds a prelimary version of the compositor's interface.

### CODE REVIEW

File: src/compositor/module.c [hunk 1]
```diff
@@ -27,3 +27,37 @@
...
+ * 
+ * This context holds the internal state of the compositor.
+ */
+static struct {
+} comp_ctx;
+
+
+int
+ws_compositor_init(void) {
+    static int is_init=0;
```
Comment (matthiasbeyer) [inline src/compositor/module.c]: Missing spaces around `=`

File: src/compositor/module.c [hunk 2]
```diff
@@ -27,3 +27,43 @@
...
+/*
+ *
+ * Interface implementation
+ *
+ */
+
+int
+ws_compositor_init(void) {
+    static int is_init = 0;
+    if (is_init) return 0;
```
Comment (matthiasbeyer) [inline src/compositor/module.c]: Brackets missing.

File: src/compositor/module.c [hunk 3]
```diff
@@ -27,3 +27,43 @@
...
+
+/*
+ *
+ * Interface implementation
+ *
+ */
+
+int
+ws_compositor_init(void) {
+    static int is_init = 0;
```
Comment (matthiasbeyer) [inline src/compositor/module.c]: `bool` would be better.

**Comment by matthiasbeyer** [comment]
Looks good to me (I don't understand that much of the code, so more documentation would be fine), but the functions you implement are to long IMHO. Can you split them up?

File: src/compositor/module.c [hunk 4]
```diff
@@ -27,3 +27,291 @@
...
+    struct ws_monitor* connector
+);
+
+/**
+ * Get all connectors on the given framebuffer device
+ *
+ * This writes directly into the given compositor context
+ */
+static int
+populate_connectors(
```
Comment (matthiasbeyer) [inline src/compositor/module.c]: Missing `void` here. Notation should be `functionname(void)` then (without newlines).

File: src/compositor/module.c [hunk 5]
```diff
@@ -27,3 +27,291 @@
...
+);
+
+
+/**
+ * Get a Frambuffer Device provided by DRM using the given path
```
Comment (matthiasbeyer) [inline src/compositor/module.c]: Typo "Frambuffer"

**Comment by TheNeikos** [comment]
The problem is that the functions are contained. Sure, I could split them up, but they would be used just once and not really benefit anything (other than having more functions).

File: src/compositor/module.c [hunk 6]
```diff
@@ -27,3 +27,291 @@
...
+);
+
+
+/**
+ * Deinitialise the compositor
+ *
+ * This function deinitialises the compositor.
+ * It does not take care of inter-module dependency nor does it check whether
+ * it was called before, unlike ws_compositor_init().
+ * This function should only be called from the main function itself.
```
Comment (matthiasbeyer) [inline src/compositor/module.c]: `@note` and `@warning` tags on the upper lines and this one would be fine.

Comment (matthiasbeyer) [inline src/compositor/module.c]: It should not be called from the main function, though. It is add to the cleaner.

File: src/compositor/module.c [hunk 7]
```diff
@@ -27,3 +27,291 @@
...
+) {
+
+    if (comp_ctx.fb.fd >= 0) {
+        close(comp_ctx.fb.fd);
+    }
+
+    struct ws_monitor* it = comp_ctx.conns;
+
+    while(it) {
+        struct ws_monitor* tmp = it->next;
```
Comment (matthiasbeyer) [inline src/compositor/module.c]: calling it `next` would be more intiuitive.

File: src/compositor/module.c [hunk 8]
```diff
@@ -27,3 +27,291 @@
...
+
+        drmModeFreeEncoder(enc);
+    }
+
+    // TODO: Log Error! No CRTC FOUND!!!
+    return -ENOENT;
+}
+
+static int
+populate_connectors(
```
Comment (matthiasbeyer) [inline src/compositor/module.c]: Missing parameter.

File: src/compositor/module.c [hunk 9]
```diff
@@ -27,3 +27,291 @@
...
+    return 0;
+}
+
+static int
+get_framebuffer_device(
+    const char* path
+) {
+    int fd = open(path, O_RDWR | O_CLOEXEC);
+    if (fd < 0)
+    {
```
Comment (matthiasbeyer) [inline src/compositor/module.c]: Should be on the line above.

File: src/compositor/module.c [hunk 10]
```diff
@@ -27,3 +27,291 @@
...
+        }
+
+        drmModeFreeEncoder(enc);
+    }
+    // There is no encoder+crtc pair! We go through /all/ the encoders now
+    for (int i = 0; i < conn->count_encoders; ++i) {
+        enc = drmModeGetEncoder(comp_ctx.fb.fd, conn->encoders[i]);
+
+        if (!enc) {
+            // TODO: Log Error!
```
Comment (matthiasbeyer) [inline src/compositor/module.c]: Notation: `@todo`

File: src/compositor/module.c [hunk 11]
```diff
@@ -27,3 +27,291 @@
...
+                drmModeFreeEncoder(enc);
+                connector->crtc = crtc;
+                return 0;
+            }
+        }
+
+        drmModeFreeEncoder(enc);
+    }
+
+    // TODO: Log Error! No CRTC FOUND!!!
```
Comment (matthiasbeyer) [inline src/compositor/module.c]: Notation: `@todo`

File: src/compositor/module.c [hunk 12]
```diff
@@ -27,3 +27,291 @@
...
+static int
+populate_connectors(
+) {
+    drmModeRes* res;
+    drmModeConnector* conn;
+    struct ws_monitor** connector = &comp_ctx.conns;
+
+    res = drmModeGetResources(comp_ctx.fb.fd);
+    if (!res) {
+        // TODO: Log Error
```
Comment (matthiasbeyer) [inline src/compositor/module.c]: Notation: `@todo`

File: src/compositor/module.c [hunk 13]
```diff
@@ -27,3 +27,291 @@
...
+        return -ENOENT;
+    }
+
+    // Let's go through all connectors (outputs)
+    int i = res->count_connectors;
+    while(i--) {
+        conn = drmModeGetConnector(comp_ctx.fb.fd, res->connectors[i]);
+        if (!conn) {
+            // TODO: Log Error
```
Comment (matthiasbeyer) [inline src/compositor/module.c]: Notation: `@todo`

File: src/compositor/module.c [hunk 14]
```diff
@@ -27,3 +27,291 @@
...
+        if (*connector) {
+            (*connector)->next = calloc(1, sizeof(**connector));
+            *connector = (*connector)->next;
+        } else {
+            *connector = calloc(1, sizeof(**connector));
+        }
+        (*connector)->conn = conn->connector_id;
+
+        if (conn->connection != DRM_MODE_CONNECTED) {
+            // TODO: Log Unused
```
Comment (matthiasbeyer) [inline src/compositor/module.c]: Notation: `@todo`

File: src/compositor/module.c [hunk 15]
```diff
@@ -27,3 +27,291 @@
...
+            (*connector)->connected = 0;
+            continue;
+        }
+
+        if (conn->count_modes == 0) {
+            // TODO: Log No Valid Modes
```
Comment (matthiasbeyer) [inline src/compositor/module.c]: Notation: `@todo`

File: src/compositor/module.c [hunk 16]
```diff
@@ -27,3 +27,291 @@
...
+            (*connector)->connected = 0;
+            continue;
+        }
+
+        // TODO: Do not just take the biggest mode available
```
Comment (matthiasbeyer) [inline src/compositor/module.c]: Notation: `@todo`

File: src/compositor/module.c [hunk 17]
```diff
@@ -27,3 +27,291 @@
...
+        memcpy(&(*connector)->mode, &conn->modes[0],
+                sizeof((*connector)->mode));
+
+        (*connector)->width = conn->modes[0].hdisplay;
+        (*connector)->height = conn->modes[0].vdisplay;
+
+        int ret = find_crtc(res, conn, *connector);
+        if (ret < 0) {
+            // TODO: Log error about not finding crtcs
```
Comment (matthiasbeyer) [inline src/compositor/module.c]: Notation: `@todo`

File: src/compositor/module.c [hunk 18]
```diff
@@ -27,3 +27,291 @@
...
+}
+
+static int
+get_framebuffer_device(
+    const char* path
+) {
+    int fd = open(path, O_RDWR | O_CLOEXEC);
+    if (fd < 0)
+    {
+        // TODO: Log Error!
```
Comment (matthiasbeyer) [inline src/compositor/module.c]: Notation: `@todo`

File: src/compositor/module.c [hunk 19]
```diff
@@ -27,3 +27,291 @@
...
+        return -ENOENT;
+    }
+
+    uint64_t has_dumb;
+    if (drmGetCap(fd, DRM_CAP_DUMB_BUFFER, &has_dumb) < 0 || !has_dumb) {
+        //TODO: Log Error!
```
Comment (matthiasbeyer) [inline src/compositor/module.c]: Notation: `@todo`

File: src/compositor/module.c [hunk 20]
```diff
@@ -27,3 +27,307 @@
...
+    struct ws_monitor* conns; //<! A linked list of ws_monitors
+} comp_ctx;
+
+/**
+ * Find the connector associated with a crtc id
+ *
+ * This function iterates over all connectors, searching for a connector with
+ * the id specified.
+ *
+ * @return -1 if a connector exists, 0 otherwise
```
Comment (matthiasbeyer) [inline src/compositor/module.c]: Why `-1` and not `true` or `false` ?

File: src/compositor/module.c [hunk 21]
```diff
@@ -27,3 +27,307 @@
...
+/**
+ * Find a suitable connector and return it's id
+ *
+ * This function tries to find a connector which is
+ *  - free
+ *  - matches the encoding
+ *  - can be used to access the monitor passed to the function
+ * and sets the crtc for a monitor
+ *
+ *  @returns 0 or a negative error value
```
Comment (matthiasbeyer) [inline src/compositor/module.c]: Leading space.

Also, can you specify which error codes? Would be nice!

File: src/compositor/module.h [hunk 22]
```diff
@@ -28,6 +28,50 @@
...
+ * A connector, which under DRM translates to a Device+Output
+ */
+struct ws_monitor {
+    struct ws_monitor* next;
+    int connected;
+    uint32_t width;
+    uint32_t height;
+    uint32_t stride;
+
+    uint32_t size;
```
Comment (matthiasbeyer) [inline src/compositor/module.h]: Size for what? Please document.

File: src/compositor/module.h [hunk 23]
```diff
@@ -28,6 +28,50 @@
...
+    uint8_t* map;
```
Comment (matthiasbeyer) [inline src/compositor/module.h]: Map for what? Please document.

File: src/compositor/module.h [hunk 24]
```diff
@@ -28,6 +28,50 @@
...
+
+    drmModeModeInfo mode;
+    uint32_t fb;
```
Comment (matthiasbeyer) [inline src/compositor/module.h]: Missing documentation here.

File: src/compositor/module.h [hunk 25]
```diff
@@ -28,6 +28,50 @@
...
+    uint32_t conn;
```
Comment (matthiasbeyer) [inline src/compositor/module.h]: Missing documentation here.

File: src/compositor/module.h [hunk 26]
```diff
@@ -28,6 +28,50 @@
...
+    uint32_t crtc;
```
Comment (matthiasbeyer) [inline src/compositor/module.h]: Missing documentation here.

File: src/compositor/module.h [hunk 27]
```diff
@@ -28,6 +28,50 @@
...
+
+/**
+ * A connector, which under DRM translates to a Device+Output
+ */
+struct ws_monitor {
+    struct ws_monitor* next;
+    int connected;
+    uint32_t width;
+    uint32_t height;
+    uint32_t stride;
```
Comment (matthiasbeyer) [inline src/compositor/module.h]: Missing documentation here.

File: src/compositor/module.h [hunk 28]
```diff
@@ -28,6 +28,50 @@
...
+    int fd; //<! The FileDescriptor
+    const char* path; //<! The path to the given filedescriptor
+};
+
+/**
+ * A connector, which under DRM translates to a Device+Output
+ */
+struct ws_monitor {
+    struct ws_monitor* next;
+    int connected;
```
Comment (matthiasbeyer) [inline src/compositor/module.h]: Why not `bool`?

File: src/compositor/module.c [hunk 29]
```diff
@@ -27,3 +27,306 @@
 
 #include "compositor/module.h"
 
+#include <errno.h>
+#include <stdbool.h>
+#include <string.h>
+#include <sys/types.h>
+#include <fcntl.h>
+#include <unistd.h>
+#include <xf86drm.h>
```
Comment (matthiasbeyer) [inline src/compositor/module.c:36]: Please sort the headers.

File: src/compositor/module.c [hunk 30]
```diff
@@ -27,3 +27,306 @@
...
+    struct ws_monitor* conns; //<! A linked list of ws_monitors
+} comp_ctx;
+
+/**
+ * Find the connector associated with a crtc id
+ *
+ * This function iterates over all connectors, searching for a connector with
+ * the id specified.
+ *
+ * @return the connector for a crtc or NULL, if not such connector exists.
```
Comment (matthiasbeyer) [inline src/compositor/module.c]: Typo "not".

File: src/compositor/module.c [hunk 31]
```diff
@@ -27,3 +27,306 @@
...
+/**
+ * Find a suitable connector and return it's id
+ *
+ * This function tries to find a connector which is
+ *  - free
+ *  - matches the encoding
+ *  - can be used to access the monitor passed to the function
+ * and sets the crtc for a monitor
+ *
+ * @returns 0 or a negative error value (-NOENT)
```
Comment (matthiasbeyer) [inline src/compositor/module.c]: The name is "ENOENT"

File: src/compositor/module.c [hunk 32]
```diff
@@ -27,3 +27,306 @@
...
+ */
+static int
+find_crtc(
+    drmModeRes* res,
+    drmModeConnector* conn,
+    struct ws_monitor* connector
+);
+
+/**
+ * Get all connectors on the given framebuffer device
```
Comment (matthiasbeyer) [inline src/compositor/module.c]: "Get" ?

File: src/compositor/module.c [hunk 33]
```diff
@@ -27,3 +27,306 @@
...
+ *
+ * This writes directly into the given compositor context
```
Comment (matthiasbeyer) [inline src/compositor/module.c]: `@return` missing.

File: src/compositor/module.c [hunk 34]
```diff
@@ -27,3 +27,306 @@
...
+int
+ws_compositor_init(void) {
+    static bool is_init = false;
+    if (is_init) {
+        return 0;
+    }
+
+    ws_cleaner_add(ws_compositor_deinit, NULL);
+
+    get_framebuffer_device("/dev/dri/card0");
```
Comment (matthiasbeyer) [inline src/compositor/module.c]: Return type ignored?

File: src/compositor/module.c [hunk 35]
```diff
@@ -27,3 +27,306 @@
...
+
+    populate_connectors();
```
Comment (matthiasbeyer) [inline src/compositor/module.c]: Return type ignored?

File: src/compositor/module.c [hunk 36]
```diff
@@ -27,3 +27,306 @@
...
+    //!< @todo: free all of the framebuffers
+
+    //!< @todo: prelimary: free the preloaded PNG
+}
+
+static struct ws_monitor*
+find_connector_with_crtc(
+        int crtc
+) {
+    for(struct ws_monitor* iter = comp_ctx.conns; iter; iter = iter->next) {
```
Comment (matthiasbeyer) [inline src/compositor/module.c]: `for(` -> `for (`

**Comment by matthiasbeyer** [comment]
Feel free to squash.

**Comment by neithernut** [comment]
Done.

**Comment by matthiasbeyer** [comment]
ACK.

**Pull request merged**